### PR TITLE
Add option to FIPS module to enforce EMS check during KDF TLS1_PRF

### LINKED ISSUE #19989: Add support for mandating use of EMS in TLS 1.2
[FIPS 140-3 Implementation Guidance](https://csrc.nist.gov/CSRC/media/Projects/cryptographic-module-validation-program/documents/fips%20140-3/FIPS%20140-3%20IG.pdf) requires that only EMS KDF in TLS 1.2 can be used with modules validated after May 16, 2023 (see "D.Q Transition of the TLS 1.2 KDF to Support the Extended Master Secret").

Please add a setting that would enable and _require_ negotiation of EMS extension for TLS 1.2 connections. That setting then could default to enabled when operating in FIPS mode.

### PR DESCRIPTION
**Pull request by slontis**
Fixes #19989

The check is disabled by default.
It can be enabled via fipsinstall using '-ems_check'
~~It can be removed from openssl config using 'no-fips-ems_check'~~

To detect its use inside the TLS_PRF it looks at the label.. If the label is 'mastersecret' then the check fails.
TLS1_PRF is used multiple times for other purposes, so if we need to pass it as a param then it may need multiple params passed.

##### Checklist

- [x] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

**Comment by t8m** [comment]
CI is relevant

File: INSTALL.md [hunk 1]
```diff
@@ -788,6 +788,11 @@ Build (and install) the FIPS provider
 Don't perform FIPS module run-time checks related to enforcement of security
 parameters such as minimum security strength of keys.
 
+### no-fips-ems_check
+
+Don't perform FIPS module run-time check related to enforcement of EMS
+during TLS1_PRF KDF.
+
```
Comment (t8m) [inline INSTALL.md:795]: Given this check is disabled by default I do not think we need a build-time configuration option for this functionality.

File: test/recipes/90-test_sslapi.t [hunk 2]
```diff
@@ -40,15 +45,91 @@ ok(run(test(["sslapitest", srctop_dir("test", "certs"),
...
+                    srctop_file("test",
+                                "recipes",
+                                "90-test_sslapi_data",
+                                "dhparams.pem")])),
+       "running sslapitest");
 }
 
 unlink $tmpfilename;
+unlink srctop_file("test", $fipsmodcfgnew);
+unlink $provconfnew;
```
Comment (t8m) [inline test/recipes/90-test_sslapi.t:135]: Add an EOL, please

File: INSTALL.md (hunk 1 above)
Comment (paulidale) [reply INSTALL.md:795]: I prefer having the option available like we do for the security checks.

Comment (t8m) [reply INSTALL.md:795]: But these are on by default. I believe this will only confuse people.

Comment (slontis) [reply INSTALL.md:795]: I would prefer to keep it consistent, they have different purposes - one strips out the code.

Comment (slontis) [reply INSTALL.md]: I have updated the doc to make it clearer.

File: apps/fipsinstall.c [hunk 3]
```diff
@@ -50,15 +51,17 @@ const OPTIONS fipsinstall_options[] = {
...
+    {"no_conditional_errors", OPT_NO_CONDITIONAL_ERRORS, '-',
+     "Disable the ability of the fips module to enter an error state if"
+     " any conditional self tests fail"},
     {"no_security_checks", OPT_NO_SECURITY_CHECKS, '-',
      "Disable the run-time FIPS security checks in the module"},
     {"self_test_onload", OPT_SELF_TEST_ONLOAD, '-',
      "Forces self tests to always run on module load"},
     {"self_test_oninstall", OPT_SELF_TEST_ONINSTALL, '-',
      "Forces self tests to run once on module installation"},
+     {"ems_check", OPT_TLS_PRF_EMS_CHECK, '-',
```
Comment (paulidale) [inline apps/fipsinstall.c:63]: indent

File: providers/implementations/kdfs/tls1_prf.c [hunk 4]
```diff
@@ -78,6 +86,8 @@ static int tls1_prf_alg(EVP_MAC_CTX *mdctx, EVP_MAC_CTX *sha1ctx,
                         unsigned char *out, size_t olen);
 
 #define TLS1_PRF_MAXBUF 1024
+#define TLS_MD_MASTER_SECRET_CONST        "\x6d\x61\x73\x74\x65\x72\x20\x73\x65\x63\x72\x65\x74"
+#define TLS_MD_MASTER_SECRET_CONST_SIZE   13
```
Comment (paulidale) [inline providers/implementations/kdfs/tls1_prf.c:90]: Use sizeof() instead of magic number?

**Review by paulidale — COMMENTED** [review]
A changes entry is probably justified.

File: providers/implementations/kdfs/tls1_prf.c (hunk 4 above)
Comment (slontis) [reply providers/implementations/kdfs/tls1_prf.c:90]: This code is a code of the SSL code..

**Review by paulidale — APPROVED** [review]

File: Configure [hunk 5]
```diff
@@ -455,6 +455,7 @@ my @disablables = (
     "filenames",
     "fips",
     "fips-securitychecks",
+    "fips-ems_check",
```
Comment (t8m) [inline Configure:458]: I still feel like this is an overkill to have this as a separate build option.

**Comment by t8m** [comment]
**OTC: do we want the FIPS EMS check as a separate build time option even if this check is off by default?**

File: Configure (hunk 5 above)
Comment (paulidale) [reply Configure:458]: I think we'll want this once the transition comes into force.

**Comment by paulidale** [comment]
OTC: ditch the configure time option.

**Comment by t-j-h** [comment]
approved subject to removal of the compile time build option as per OTC disucssion

**Comment by slontis** [comment]
build time option removed

**Review by paulidale — APPROVED** [review]
still good

**Comment by slontis** [comment]
Rebased to fix a conflict, and removed spacing change in Configure file  (So Configure is not part of the commit)

**Comment by t8m** [comment]
CI is relevant

File: doc/man7/OSSL_PROVIDER-FIPS.pod [hunk 6]
```diff
@@ -189,6 +204,21 @@ TEST-RAND is an unapproved algorithm.
...
+
+=over 4
+
+=item "security-checks" (B<OSSL_OSSL_PROV_PARAM_SECURITY_CHECKS>) <unsigned integer>
+
+For further information refer to the L<openssl-fipsinstall(1)> option
+B<-no_security_checks>.
+
+=back
+
```
Comment (t8m) [inline doc/man7/OSSL_PROVIDER-FIPS.pod:221]: Why this duplication?

**Review by t8m — CHANGES_REQUESTED** [review]

File: doc/man7/OSSL_PROVIDER-FIPS.pod (hunk 6 above)
Comment (slontis) [reply doc/man7/OSSL_PROVIDER-FIPS.pod]: Merge error.... Fixed.
Added a CHANGES entry also.

**Review by paulidale — APPROVED** [review]

**Comment by slontis** [comment]
@paulidale? Did you want this in 3.1? Waiting till OTC meeting put this inside the freeze period?

**Comment by paulidale** [comment]
Yes to 3.1.  We will need this come the transitions.

**Comment by paulidale** [comment]
OTC: **should this go to 3.1?**

**Comment by paulidale** [comment]
OTC: this is okay to go into 3.1.

**Comment by t8m** [comment]
Merged to master and 3.1 branches. Thank you.

**Pull request closed without merging**